### PR TITLE
[Tiered Storage] Exclude current ledger from offload deque

### PR DESCRIPTION
**Pull request by gaoran10**
### Motivation

Currently, the current ledger maybe adds to the offload deque.

If the method `ManagedLedgerImpl#maybeOffload` try to acquire the lock `offloadMutex` failed, the offload operation will be scheduled executed after 100 ms in the background, and the new ledger may be created successfully, then the retry offload operation will add the current new ledger to offload queue, because the current ledger is incomplete and the offload data will be incomplete.

### Modifications

Exclude current ledger from offload deque.

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (no)
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: (no)

### CODE REVIEW

**Review by sijie — APPROVED** [review]
@gaoran10 Can you add a test case?

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java [hunk 1]
```diff
@@ -1411,9 +1411,21 @@ synchronized void ledgerClosed(final LedgerHandle lh) {
             log.debug("[{}] Ledger has been closed id={} entries={}", name, lh.getId(), entriesInLedger);
         }
         if (entriesInLedger > 0) {
-            LedgerInfo info = LedgerInfo.newBuilder().setLedgerId(lh.getId()).setEntries(entriesInLedger)
-                    .setSize(lh.getLength()).setTimestamp(clock.millis()).build();
-            ledgers.put(lh.getId(), info);
+            synchronized (ledgers) {
```
Comment (eolivelli) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:1414]: We are not using synchronized on ledgers variable in other points in this class.
It looks to me as an inconsistent synchronisation.

**Review by eolivelli — CHANGES_REQUESTED** [review]

**Comment by wolfstudy** [comment]
@gaoran10 can you check @eolivelli comments? thanks

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java (hunk 1 above)
Comment (gaoran10) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java]: Thanks, eolivelli, I think you are right.

I checked the code and found out all situations that change the `ledgers` in the `ManagedLedgerImpl`, for a ledger its `LedgerInfo` update operations don't exist competition.

    1. When the ManagedLedgerImpl initialized, recover metadata from the metadata store.
    2. When creating a new ledger.
    3. When the ledger closed.
    4. When the offloaded ledger data should be deleted from bk.
    5. When preparing for offload.
    6. When offloading complete.

I think the `LedgerInfo` should be initialized when the ledger created, after this, the `LedgerInfo` should be updated not be created.

PTAL

**Comment by gaoran10** [comment]
@sijie Ok, I'll try to add a test, but it's not easy to cover the problem.

**Review by codelipenghui — APPROVED** [review]

**Comment by wolfstudy** [comment]
@gaoran10 Please check the action CI, thanks.

**Comment by gaoran10** [comment]
The current ledge would not be added to the offload queue, I close this PR first.

**Pull request closed without merging**